homeassistant/components/history/__init__.py

```diff
@@ -300,13 +362,401 @@ def _sorted_significant_states_json(
+        state.last_updated
+    )
+    if state.last_changed != state.last_updated:
+        comp_state[COMPRESSED_STATE_LAST_CHANGED] = dt_util.utc_to_timestamp(
+            state.last_changed
+        )
+    return comp_state
+
+
+def _events_to_compressed_states(
+    events: Iterable[Event], no_attributes: bool
+) -> MutableMapping[str, list[dict[str, Any]]]:
+    """Convert events to a compressed states."""
+    states_by_entity_ids: dict[str, list[dict[str, Any]]] = {}
+    for event in events:
+        state: State | None = event.data.get("new_state")
+        if state is None:
+            continue
+        entity_id: str = state.entity_id
+        states_by_entity_ids.setdefault(entity_id, []).append(
+            _history_compressed_state(state, no_attributes)
+        )
+    return states_by_entity_ids
+
+
+async def _async_events_consumer(
+    subscriptions_setup_complete_time: dt,
+    connection: ActiveConnection,
+    msg_id: int,
+    stream_queue: asyncio.Queue[Event],
+    no_attributes: bool,
+) -> None:
+    """Stream events from the queue."""
+    while True:
+        events: list[Event] = [await stream_queue.get()]
+        # If the event is older than the last db
+        # event we already sent it so we skip it.
+        if events[0].time_fired <= subscriptions_setup_complete_time:
+            continue
+        # We sleep for the EVENT_COALESCE_TIME so
+        # we can group events together to minimize
+        # the number of websocket messages when the
+        # system is overloaded with an event storm
+        await asyncio.sleep(EVENT_COALESCE_TIME)
+        while not stream_queue.empty():
+            events.append(stream_queue.get_nowait())
+
+        if history_states := _events_to_compressed_states(events, no_attributes):
+            connection.send_message(
+                JSON_DUMP(
+                    messages.event_message(
+                        msg_id,
+                        {"states": history_states},
+                    )
+                )
+            )
+
+
+@callback
+def _async_subscribe_events(
+    hass: HomeAssistant,
+    subscriptions: list[CALLBACK_TYPE],
+    target: Callable[[Event], None],
+    entities_filter: EntityFilter | None,
+    entity_ids: list[str] | None,
+    significant_changes_only: bool,
+    minimal_response: bool,
+) -> None:
+    """Subscribe to events for the entities and devices or all.
+
+    These are the events we need to listen for to do
+    the live history stream.
+    """
+    assert is_callback(target), "target must be a callback"
+
+    @callback
+    def _forward_state_events_filtered(event: Event) -> None:
+        """Filter state events and forward them."""
+        if (new_state := event.data.get("new_state")) is None:
+            return
+        if entities_filter and not entities_filter(new_state.entity_id):
+            return
+        if not significant_changes_only and not minimal_response:
```

Comment:
Unfortunately this logic needs the same as https://github.com/home-assistant/core/blob/77f9548e51fa6bd3df3b0d864fb5c587ac2b18e1/homeassistant/components/recorder/history.py#L281

Reply:
SIGNIFICANT_DOMAINS